eth/tracers/api.go

```diff
@@ -862,71 +862,78 @@ func (api *API) TraceCall(ctx context.Context, args ethapi.TransactionArgs, bloc
 // executes the given message in the provided environment. The return value will
 // be tracer dependent.
 func (api *API) traceTx(ctx context.Context, message core.Message, txctx *Context, vmctx vm.BlockContext, statedb *state.StateDB, config *TraceConfig) (interface{}, error) {
-	// Assemble the structured logger or the JavaScript tracer
 	var (
-		tracer    vm.EVMLogger
 		err       error
 		txContext = core.NewEVMTxContext(message)
 	)
-	switch {
-	case config == nil:
-		tracer = logger.NewStructLogger(nil)
-	case config.Tracer != nil:
-		// Define a meaningful timeout of a single transaction trace
-		timeout := defaultTraceTimeout
-		if config.Timeout != nil {
-			if timeout, err = time.ParseDuration(*config.Timeout); err != nil {
-				return nil, err
-			}
-		}
-		if t, err := New(*config.Tracer, txctx); err != nil {
+	// Use struct logger
+	if config == nil || config.Tracer == nil {
+		return api.logTx(message, txctx, vmctx, statedb, config, txContext)
+	}
+
+	// Define a meaningful timeout of a single transaction trace
+	timeout := defaultTraceTimeout
+	if config.Timeout != nil {
+		if timeout, err = time.ParseDuration(*config.Timeout); err != nil {
 			return nil, err
-		} else {
-			deadlineCtx, cancel := context.WithTimeout(ctx, timeout)
-			go func() {
-				<-deadlineCtx.Done()
-				if errors.Is(deadlineCtx.Err(), context.DeadlineExceeded) {
-					t.Stop(errors.New("execution timeout"))
-				}
-			}()
-			defer cancel()
-			tracer = t
 		}
-	default:
-		tracer = logger.NewStructLogger(config.Config)
 	}
+	tracer, err := New(*config.Tracer, txctx)
+	if err != nil {
+		return nil, err
+	} else {
+		deadlineCtx, cancel := context.WithTimeout(ctx, timeout)
+		go func() {
+			<-deadlineCtx.Done()
+			if errors.Is(deadlineCtx.Err(), context.DeadlineExceeded) {
+				tracer.Stop(errors.New("execution timeout"))
+			}
+		}()
+		defer cancel()
+	}
+
```

Comment:
rm

Reply:
Do you mean remove `defer cancel()`? I checked out the docs out of curiosity, it says it releases the context resources if the operation ends before the timeout. Isn't that kinda nice to keep?

Reply:
No I just meant to drop the empty line after ending bracket (but bracket is gone now, so the empty line can stay)